rusoto/core/src/signature.rs

```diff
@@ -184,6 +184,94 @@ impl SignedRequest {
         self.params = params;
     }
 
+
+    /// The request has to be signed *before* calling presigned_url
+    /// http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+    pub fn generate_presigned_url(&mut self, creds: &AwsCredentials) -> String {
+        debug!("Presigning request URL");
+
+        self.sign(creds);
+        let hostname = match self.hostname {
+            Some(ref h) => h.to_string(),
+            None => build_hostname(&self.service, &self.region),
+        };
+
+        let current_time = now_utc();
+        let current_time_fmted = current_time.strftime("%Y%m%dT%H%M%SZ").unwrap();
+        let current_date = current_time.strftime("%Y%m%d").unwrap();
+
+        let signed_headers = signed_headers(&self.headers);
+        self.canonical_uri = canonical_uri(&self.path);
+
+        self.remove_header("x-amz-content-sha256");
+        self.add_header("X-Amz-Content-Sha256", "");
+        self.remove_header("X-Amz-Date");
+        self.add_header("X-Amz-Date", &format!("{}", &current_time_fmted));
+        self.params.put("X-Amz-Date", format!("{}", &current_time_fmted));
+        self.params.put("x-amz-date", format!("{}", &current_time_fmted));
+
+        self.params.put("x-amz-content-sha256", "");
+        self.remove_header("content-type");
+        self.add_header("content-type", "");
+
+        self.params.put("X-Amz-Algorithm", "AWS4-HMAC-SHA256");
+        self.params.put("X-Amz-Credential", format!("{}/{}/{}/{}/aws4_request", &creds.aws_access_key_id(), format!("{}", &current_date), self.region, self.service));
```

Comment:
I believe this change should fix Ceph support:

`self.region` → `self.region.name()`

The `Display` implementation for `Region` converts this to `"us-east-1 with endpoint http://localhost:9000"`. Not sure if the Display implementation should be changed or even removed. It's the second time I've seen this used the wrong way (3033dea24a6cc23c8072d5f1baae0e6e9fc31162).

Reply:
I decided to just create a pull request for removing the `Display` impl (#1040 ).